core/Pos/Core/Slotting.hs

```diff
@@ -71,6 +79,14 @@ crucialSlot epochIdx = SlotId {siEpoch = epochIdx - 1, ..}
         leftToPanic "crucialSlot: " $
         mkLocalSlotIndex (fromIntegral (epochSlots - slotSecurityParam - 1))
 
+instance HasEpochIndex EpochOrSlot where
+    epochIndexL = lens (epochOrSlot identity siEpoch) setter
```

Comment:
There's a `HasEpochIndex` instance for `Either`, its implementation is way simpler because it uses the `choosing` combinator. Since `EpochOrSlot` is a newtype over `Either Epoch Slot`, I think you'd want to reuse that instance, perhaps like this:

```
epochIndexL = _EpochOrSlot . epochIndexL
```

where `_EpochOrSlot` is an isomorphism that can be generated by `makePrisms ''EpochOrSlot`.